### PR TITLE
[READY]Gives AI a shortcut to open the nearest door to someone.

### PR DESCRIPTION
**Pull request by silicons**
Requested by @YakumoChen 

Uses the same things as tracking to ensure they're on cameras, pretty sure this works.
Feedbacks with the name of the door that they open.
:cl:
rscadd: AIs now get a link to open the nearest door to someone on radio!
/:cl:

WIP

### CODE REVIEW

**Comment by MrPerson** [comment]
Dude no you've gone too far

**Comment by ShizCalev** [comment]
I love you.

Does this check if the door is visible via camera as well?

**Comment by kevinz000** [comment]
^ Shit, I'll do that.

**Comment by YakumoChen** [comment]
I told Mekhi to code the "AI DOOR!!!" button... he actually did it the absolute madman hahahahahaha!!!!!!!

**Comment by jughu** [comment]
This cant end well 

Inb4 guy asks ai to open door then quickly run to another door and that door opens this is the best pr of trickery

**Comment by XDTM** [comment]
It's already there on goon, as far as i know. As long as the area is displayed to the ai to avoid armory raids i think it's fine, and it also adds the emergent gameplay of renaming secure areas into harmless ones with blueprints so the AI doesn't check on you.

**Comment by kevinz000** [comment]
right trying to get hrefs to work.

**Comment by ShadowDeath6** [comment]
Paradise prompts AI saying "do you want to open [door.name] for person?" When you hit the open button. That could stop some of the cheese of going to a different door before the AI does it.

**Comment by xxalpha** [comment]
This sounds like it's going to be abused.

**Comment by GunHog** [comment]
Include the airlock's name or at least the area in which the airlock resides. On that subject, please keep the names consistent; Choose only "Airlock" or "Door", not both.

**Comment by PKPenguin321** [comment]
No, don't. If you're too lazy to look at the person you're opening a door for, you should have to risk possibly opening the armory or whatever else they may be asking for. This could actually be a good feature that way.

**Comment by Shapsy** [comment]
There is absolutely no way in hell that it is too much effort for the AI to take two seconds to look at the door and judge whether or not opening the door is a good idea or not

**Comment by GunHog** [comment]
Very well. I rarely agree with anything you say regarding silicons for obvious reasons, PKP, but I like your point here. It has potential for LOL. LOL should not be denied!

**Comment by ShizCalev** [comment]
Imo, having "Opening [D.name] for [M.name]!" as it is already is more than enough. 

It offers the chance for the AI to screw up and accidentally allow someone where they shouldn't be, while also giving you enough notice that you've fucked up so you can inform security of how you're a bad AI and that they need to fix your terrible mistake of letting greysuit Bob into the armory.

**Comment by coiax** [comment]
This can only have hilarious implications, and that's a GOOD THING.

**Comment by BeeSting12** [comment]
so the ai cant look at the cameras itself to decide if its a good idea?

**Comment by ShizCalev** [comment]
It can. You would click the target's name as normal to jump your camera to them.

**Comment by ExcessiveUseOfCobblestone** [comment]
What's the point of having this except for demands of AI jobbans because the AI didn't take the few seconds to check whether or not the area is harmful?

It's not a QoL if it cuts corners that can cause you [the player] to be in trouble on an OOC level.

**Comment by ShizCalev** [comment]
Does this check if the AI is conscious? Ie; if the AI loses power, can it still open the door with the href?

**Comment by kevinz000** [comment]
^ noted
@ExcessiveUseOfCobblestone im adding either the door name or the area name, and if ais don't look at that...... yeah I'm still working on making the href work and then I can add in checks for ai health/door vision/do stuff to make it less confusing I guess

**Comment by PKPenguin321** [comment]
>im adding either the door name or the area name, 

don't. As someone who is also an admin, we have the ability to discern if this kind of thing is an accident or not. We're not robots. Adding an area name or whatever the fuck just turns this potentially entertaining tool for misdirection and conflict to another AI powercreep feature.

**Comment by kevinz000** [comment]
hmmmm

**Comment by coiax** [comment]
> AI powercreep

Lets the AI open doors for people without having to stop whatever it's  doing.

...

POWERCREEP.

**Comment by coiax** [comment]
@kevinz000 just have it give an OPEN link if the user says door or open, and state the area that they're in. That should fix a bunch of problems.

**Comment by PKPenguin321** [comment]
Yes it is powercreep, previously the AI would have to detract it's attention from whatever potential murder or whatever that it happens to be stalking so it can open the door. I won't have some silicon play wave it off as "xD le quality of life". Not to mention the potential for conflict and misdirection is pretty good without the area name, and it would be a shame to waste that.

**Comment by kevinz000** [comment]
I'm going to add in sanity/consciousness/whatever checking and make it working without the area/door name stating and hopefully get someone to testmerge it.

**Comment by RemieRichards** [comment]
@kevinz000 Stop with the terrible commit messages.

**Comment by kevinz000** [comment]
Sorry.

File: code/modules/mob/living/silicon/ai/ai.dm [hunk 1]
```diff
@@ -425,24 +425,37 @@ var/list/ai_list = list()
...
 	var/search_range = 0
-	for(search_range = 0, search_range < 5, search_range++)
-		for(var/obj/machinery/door/airlock/A in range(search_range,M))
+	for(search_range = 0, search_range < 6, search_range++)
+		world << "DEBUG: SEARCHING: RANGE AT [search_range]!"
+		for(var/obj/machinery/door/airlock/A in range(search_range,get_turf(M)))
+			if(!cameranet.checkTurfVis(get_turf(A)))
+				world << "DEBUG: [A] IS NOT ON CAMERAS! SKIPPING!"
+				continue
+			world << "DEBUG: FOUND [A]!"
```
Comment (PKPenguin321) [inline code/modules/mob/living/silicon/ai/ai.dm]: Lots of debug stuff left in

**Comment by PKPenguin321** [comment]
Oh what now it's gone
Weird, my bad

**Comment by YakumoChen** [comment]
the only gripe I have is that the OPEN DOOR button is kind of too wordy, maybe just make it [DOOR] or even [D]?

**Comment by YakumoChen** [comment]
Also it would be nice if I got a prompt into what door I am opening...

**Comment by GunHog** [comment]
You really should pick one word: "Airlock" or "Door", not both!

**Comment by kevinz000** [comment]
hm

**Comment by kevinz000** [comment]
@PKPenguin321 what do you think if I made it say the airlock's name AFTER it's opened but not before?

**Comment by PKPenguin321** [comment]
At that point it's probably too late anyways so go for it, would probably make for some funny moments with AIs realizing what they've done and scrambling to fix it

**Comment by kevinz000** [comment]
k

**Comment by kevinz000** [comment]
k all done :^)

**Comment by ShadowDeath6** [comment]
well if it doesn't tell you what door it is why does this PR even exist? You're punishing the AIs for using a feature that's less efficient and more exploitable by the crew?

**Comment by PKPenguin321** [comment]
Yes. If you use the quicker option without bothering to check what door you're opening, you might get punished. If you don't like it, just observe them and open the door manually like normal.

**Comment by ShadowDeath6** [comment]
That just begs the same question, "why does this pr exist?"

**Comment by ShizCalev** [comment]
^ Since I'm a tad drunk right now, I realized that very same thing two Kevinsz posts ago. Perhaps incidents involving the button should be ruled as an IC issue unless it becomes commonplace with an individual?

**Comment by ExcessiveUseOfCobblestone** [comment]
or just not add the feature when it just serves as bwoink-bait...

It's """powercreep""" with the area indicator, it's inherently dumb [harm through inaction] without.

If you cannot be arsed with checking the door, don't play the job thats default lawset makes it mainly a door opener [that also needs to make sure the individual won't be harmed because of it].

And yes, I am not a robot, hence why I am not inclined to like a """QoL"""  feature that makes people bwoink in bulk everytime the AI opens a "bad" door.

**Comment by PKPenguin321** [comment]
Not even bwoink bait.
"AI, please open this door to medbay!"
AI is tracking a known harmful criminal, so it opens the door via this button. This way it can prevent law one infractions by following the criminal while still obeying a law two order.
But it's actually the vault/bridge/brig etc., not the medbay! The AI has been bamboozled!

Would you honestly ban the AI for this, cobby?

**Comment by optimumtact** [comment]
I dislike the AI being the crew door opener and this is pushing it further in that direction, so I'm going to deny this.

**Comment by kevinz000** [comment]
:^(

**Pull request closed without merging**